File: frontend/src/lib/types/timeline.ts
/**
 * Timeline types – mirrors the OpenAPI schema from
 * backend/app/schemas/timeline.py.
 *
 * Keep in sync: any change to the Pydantic models must be reflected here.
 */

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export type StageType =
  | "literature_review"
  | "methodology"
  | "data_collection"
  | "analysis"
  | "writing"
  | "defense_preparation"
  | "revision"
  | "publication"
  | "coursework"
  | "fieldwork"
  | "other";

export type StageStatus =
  | "not_started"
  | "in_progress"
  | "completed"
  | "delayed";

export type DeliverableType =
  | "paper"
  | "presentation"
  | "dataset"
  | "code"
  | "thesis_chapter"
  | "proposal"
  | "review"
  | "report"
  | "defense"
  | "other";

export type MilestoneState =
  | "upcoming"
  | "active"
  | "completed"
  | "delayed";

export type SuggestionStatus = "pending" | "accepted" | "rejected";

export type SuggestionReason =
  | "milestone_delay"
  | "supervision_inactivity"
  | "writing_stagnation";

// ---------------------------------------------------------------------------
// Core domain objects
// ---------------------------------------------------------------------------

/** A major phase in a PhD timeline. */
export interface TimelineStage {
  id: string;
  title: string;
  description: string | null;
  stage_type: StageType | null;
  stage_order: number;
  start_date: string | null;
  end_date: string | null;
  duration_months: number | null;
  status: StageStatus;
  notes: string | null;
  /** AI confidence score 0–100. */
  confidence: number | null;
}

/** A concrete deliverable / checkpoint within a stage. */
export interface TimelineMilestone {
  id: string;
  title: string;
  description: string | null;
  deliverable_type: DeliverableType | null;
  is_critical: boolean;
  is_completed: boolean;
  milestone_order: number;
  target_date: string | null;
  actual_completion_date: string | null;
  state: MilestoneState;
  notes: string | null;
  /** Parent stage ID. */
  stage_id: string;
}

/** Directed dependency edge between stages / milestones. */
export interface Dependency {
  from: string;
  to: string;
}

/** Per-stage duration estimate. */
export interface DurationEstimate {
  stage_id: string;
  estimated_months: number | null;
  start_date: string | null;
  end_date: string | null;
}

// ---------------------------------------------------------------------------
// API response contracts
// ---------------------------------------------------------------------------

/** Full timeline payload – the canonical response. */
export interface TimelineResponse {
  timeline_id: string;
  baseline_id: string;
  title: string;
  description: string | null;
  stages: TimelineStage[];
  milestones: TimelineMilestone[];
  dependencies: Dependency[];
  durations: DurationEstimate[];
  created_at: string | null;
  updated_at: string | null;
}

/** Response after generating a draft timeline. */
export interface GenerateTimelineResponse {
  timeline_id: string;
  baseline_id: string;
  draft_timeline_id: string;
}

/** A timeline adjustment suggestion. */
export interface TimelineSuggestion {
  id: string;
  committed_timeline_id: string;
  reason: SuggestionReason;
  title: string;
  message: string;
  suggestion_payload: Record<string, unknown>;
  status: SuggestionStatus;
  responded_at: string | null;
}

File: backend/app/schemas/timeline.py
"""
Timeline Pydantic schemas.

These models define the OpenAPI contract for all timeline endpoints.
The frontend TypeScript types in lib/types/timeline.ts must stay in sync
with these definitions.
"""
from __future__ import annotations

from datetime import date, datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Enums – exposed in the OpenAPI schema for frontend codegen
# ---------------------------------------------------------------------------

class StageType(str, Enum):
    """Classification of a timeline stage."""

    LITERATURE_REVIEW = "literature_review"
    METHODOLOGY = "methodology"
    DATA_COLLECTION = "data_collection"
    ANALYSIS = "analysis"
    WRITING = "writing"
    DEFENSE_PREPARATION = "defense_preparation"
    REVISION = "revision"
    PUBLICATION = "publication"
    COURSEWORK = "coursework"
    FIELDWORK = "fieldwork"
    OTHER = "other"


class StageStatus(str, Enum):
    """Current progress status of a stage."""

    NOT_STARTED = "not_started"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    DELAYED = "delayed"


class DeliverableType(str, Enum):
    """Type of deliverable associated with a milestone."""

    PAPER = "paper"
    PRESENTATION = "presentation"
    DATASET = "dataset"
    CODE = "code"
    THESIS_CHAPTER = "thesis_chapter"
    PROPOSAL = "proposal"
    REVIEW = "review"
    REPORT = "report"
    DEFENSE = "defense"
    OTHER = "other"


class MilestoneState(str, Enum):
    """State-machine state for a milestone."""

    UPCOMING = "upcoming"
    ACTIVE = "active"
    COMPLETED = "completed"
    DELAYED = "delayed"


class SuggestionStatus(str, Enum):
    """Status of a timeline adjustment suggestion."""

    PENDING = "pending"
    ACCEPTED = "accepted"
    REJECTED = "rejected"


class SuggestionReason(str, Enum):
    """Trigger reason for an adjustment suggestion."""

    MILESTONE_DELAY = "milestone_delay"
    SUPERVISION_INACTIVITY = "supervision_inactivity"
    WRITING_STAGNATION = "writing_stagnation"


# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------

class GenerateTimelineRequest(BaseModel):
    baseline_id: UUID
    title: Optional[str] = None


class UpdateMilestoneRequest(BaseModel):
    is_completed: bool


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------

class DurationEstimate(BaseModel):
    """Per-stage duration estimate (planned vs actual)."""

    stage_id: UUID
    estimated_months: Optional[int] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None


class Dependency(BaseModel):
    """Directed dependency edge between two stages or milestones."""

    from_id: str = Field(..., alias="from")
    to_id: str = Field(..., alias="to")

    model_config = ConfigDict(populate_by_name=True)


class TimelineStageOut(BaseModel):
    """A single stage within a timeline."""

    id: UUID
    title: str
    description: Optional[str] = None
    stage_type: Optional[StageType] = None
    stage_order: int
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    duration_months: Optional[int] = None
    status: StageStatus = StageStatus.NOT_STARTED
    notes: Optional[str] = None
    confidence: Optional[float] = Field(
        None, ge=0, le=100, description="AI confidence score 0–100"
    )

    model_config = ConfigDict(from_attributes=True)


class TimelineMilestoneOut(BaseModel):
    """A single milestone within a stage."""

    id: UUID
    title: str
    description: Optional[str] = None
    deliverable_type: Optional[DeliverableType] = None
    is_critical: bool = False
    is_completed: bool = False
    milestone_order: int
    target_date: Optional[date] = None
    actual_completion_date: Optional[date] = None
    state: MilestoneState = MilestoneState.UPCOMING
    notes: Optional[str] = None
    stage_id: UUID = Field(..., description="Parent stage ID")

    model_config = ConfigDict(from_attributes=True)


class TimelineResponse(BaseModel):
    """
    Full timeline payload – the canonical response contract.

    Frontend ``TimelineResponse`` in ``lib/types/timeline.ts`` mirrors this.
    """

    timeline_id: UUID
    baseline_id: UUID
    title: str
    description: Optional[str] = None
    stages: list[TimelineStageOut]
    milestones: list[TimelineMilestoneOut]
    dependencies: list[Dependency] = []
    durations: list[DurationEstimate] = []
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class GenerateTimelineResponse(BaseModel):
    """Response after generating a draft timeline."""

    timeline_id: UUID
    baseline_id: UUID
    draft_timeline_id: UUID


class TimelineSuggestionOut(BaseModel):
    """A timeline adjustment suggestion."""

    id: UUID
    committed_timeline_id: UUID
    reason: SuggestionReason
    title: str
    message: str
    suggestion_payload: dict
    status: SuggestionStatus
    responded_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)
